src/jvm/main/org/jetbrains/kotlinx/lincheck/ObjectTraverser.kt

```diff
@@ -0,0 +1,190 @@
+    createObjectToNumberMap(obj, Collections.newSetFromMap(IdentityHashMap()), objectNumberMap)
+
+    return objectNumberMap
+}
+
+/**
+ * Recursively traverses an object to enumerate it and all nested objects.
+ *
+ * @param obj object to traverse
+ * @param processedObjects a set of already processed objects. Required in case of cyclic references.
+ * @param objectNumberMap result enumeration map
+ */
+private fun createObjectToNumberMap(obj: Any, processedObjects: MutableSet<Any>, objectNumberMap: MutableMap<Any, Int>) {
+    if (!processedObjects.add(obj)) return
+    val objectNumber = getObjectNumber(obj.javaClass, obj)
+    objectNumberMap[obj] = objectNumber
+
+    var clazz: Class<*>? = obj.javaClass
+    if (clazz!!.isArray) {
+        if (obj is Array<*>) {
+            obj.forEach { element -> element?.let { createObjectToNumberMap(it, processedObjects, objectNumberMap) } }
+        }
+        return
+    }
+    while (clazz != null) {
+        clazz.declaredFields.filter { f ->
+            !Modifier.isStatic(f.modifiers) &&
+                    !f.isEnumConstant &&
+                    f.name != "serialVersionUID"
+        }.forEach { f ->
+            try {
+                var value: Any? = readField(obj, f)
+
+                if (isAtomic(value)) {
+                    value = value!!.javaClass.getDeclaredMethod("get").invoke(value)
+                }
+
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicRef") {
+                    value = readField(value, value.javaClass.getDeclaredField("value"))
+                }
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicInt") {
+                    value = readField(value, value.javaClass.getDeclaredField("value"))
+                }
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicLong") {
+                    value = readField(value, value.javaClass.getDeclaredField("value"))
+                }
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicBoolean") {
+                    value = readField(value, value.javaClass.getDeclaredField("_value"))
+                }
+
+                if (value is AtomicIntegerArray) {
+                    value = (0..value.length()).map { (value as AtomicIntegerArray).get(it) }.toIntArray()
+                }
+                if (value is AtomicReferenceArray<*>) {
+                    value = (0..value.length()).map { (value as AtomicReferenceArray<*>).get(it) }.toTypedArray()
+                }
+
+                if (value?.javaClass?.canonicalName?.startsWith("java.lang.invoke.") == true) {
+                    // Ignore
+                } else if (value is AtomicReferenceFieldUpdater<*, *> || value is AtomicIntegerFieldUpdater<*> || value is AtomicLongFieldUpdater<*>) {
+                    // Ignore
+                } else if (value is ReentrantLock) {
+                    // Ignore
```

Comment:
Why not `ReentrantLock(isLocked = <isLocked>)`?

Reply:
This method only enumerates objects, not creating their string representation

Reply:
Why don't you need to enumerate ReentrantLock?

Reply:
Ok, there is no reason to ignore it. Removed this if